nova_vm/src/engine/bytecode/executable.rs

```diff
@@ -2040,74 +2040,224 @@ fn complex_array_pattern(
-                ));
+    if let Some(rest) = &pattern.rest {
+        match &rest.argument.kind {
+            ast::BindingPatternKind::BindingIdentifier(identifier) => {
                 let identifier_string = ctx.create_identifier(&identifier.name);
                 ctx.exe.add_instruction_with_identifier(
-                    Instruction::BindingPatternBind,
+                    Instruction::BindingPatternBindRest,
                     identifier_string,
                 );
-            } else {
-                let key_string = match &ele.key {
-                    ast::PropertyKey::StaticIdentifier(identifier) => {
-                        String::from_str(ctx.agent, &identifier.name).into_value()
-                    }
-                    ast::PropertyKey::NumericLiteral(literal) => {
-                        let numeric_value = Number::from_f64(ctx.agent, literal.value);
-                        if let Number::Integer(_) = numeric_value {
-                            numeric_value.into()
-                        } else {
-                            Number::to_string_radix_10(ctx.agent, numeric_value).into_value()
-                        }
-                    }
-                    ast::PropertyKey::StringLiteral(literal) => {
-                        String::from_str(ctx.agent, &literal.value).into_value()
-                    }
-                    _ => todo!("Complex object patterns"),
-                };
-
-                match &ele.value.kind {
-                    ast::BindingPatternKind::BindingIdentifier(identifier) => {
-                        let value_identifier_string = ctx.create_identifier(&identifier.name);
-                        ctx.exe.add_instruction_with_identifier_and_constant(
-                            Instruction::BindingPatternBindNamed,
-                            value_identifier_string,
-                            key_string,
-                        )
+            }
+            _ => todo!("Complex object patterns"),
+        }
+    } else {
+        ctx.exe.add_instruction(Instruction::FinishBindingPattern);
+    }
+}
+
+fn complex_object_pattern(
+    object_pattern: &ast::ObjectPattern<'_>,
+    ctx: &mut CompileContext,
+    has_environment: bool,
+) {
+    // 8.6.2 Runtime Semantics: BindingInitialization
+    // BindingPattern : ObjectBindingPattern
+    // 1. Perform ? RequireObjectCoercible(value).
+    // NOTE: RequireObjectCoercible throws in the same cases as ToObject, and other operations
+    // later on (such as GetV) also perform ToObject, so we convert to an object early.
+    ctx.exe.add_instruction(Instruction::Store);
+    ctx.exe.add_instruction(Instruction::ToObject);
+    ctx.exe.add_instruction(Instruction::Load);
+
+    for property in &object_pattern.properties {
+        match &property.key {
+            ast::PropertyKey::StaticIdentifier(identifier) => {
+                ctx.exe.add_instruction(Instruction::Store);
```

Comment:
**thought:** Do we need a `StoreCopy` as well? :)

Reply:
Maybe at some point.